### PR TITLE
New vector field scheme

### PR DESCRIPTION
**Pull request by mfschubert**
Introduce new vector field generating schemes, which are similar to existing but suffixed with "_FOURIER". These are faster while offering similar vector fields. It addresses #28, and the new schemes give proper vector fields for odd-shaped unit cells (see issue #55).

Speed comparison:
```
shape = (45, 120), approximate_num_terms = 200
  Fourier scheme:  25.3 ms ± 1.68 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
  existing scheme: 908 ms ± 2.68 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
shape = (45, 120), approximate_num_terms = 400
  Fourier scheme:  63.6 ms ± 3.16 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
  existing scheme: 907 ms ± 1.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
shape = (140, 140), approximate_num_terms = 800
  Fourier scheme:  334 ms ± 11.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
  existing scheme: 3.86 s ± 1.75 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
shape = (140, 140), approximate_num_terms = 2000
  Fourier scheme:  1.28 s ± 55.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
  existing scheme: 3.87 s ± 22.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

uLED example convergence of extraction efficiency
image

Vector field image produced by vector_fields.py example
vector_fields

In the future we should consider removing the original schemes.

### CODE REVIEW

File: fmmax/vector_fourier.py [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+    primitive_lattice_vectors: basis.LatticeVectors,
+    use_jones_direct: bool,
+    fourier_loss_weight: float,
+    steps: int = 1,
+) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    """Compute the tangent vector field for `arr`.
+
+    The calculation finds the minimum of a quadratic loss function using a single
+    Newton iteration. Rather than optimizing the real-space tangent field, the
+    Fourier coefficients are directly optimized.
```
Comment (smartalecH) [inline fmmax/vector_fourier.py:101]: This is the main innovation to these approaches, right?

Comment (mfschubert) [reply fmmax/vector_fourier.py:101]: Essentially, yes. More explicitly, the change is is a (1) a new loss function, (2) a new parameterization (solving for the Fourier coefficients directly, rather than the vector field on the real-space grid), and (3) the use of Newton method rather than a first order method to minimize the loss. The update to fix vector fields for odd-shaped unit cells is a further change.

File: fmmax/fmm.py [hunk 2]
```diff
@@ -28,6 +28,10 @@ class Formulation(enum.Enum):
     JONES: str = vector.JONES
     NORMAL: str = vector.NORMAL
     POL: str = vector.POL
+    JONES_DIRECT_FOURIER: str = vector.JONES_DIRECT_FOURIER
+    JONES_FOURIER: str = vector.JONES_FOURIER
+    NORMAL_FOURIER: str = vector.NORMAL_FOURIER
+    POL_FOURIER: str = vector.POL_FOURIER
```
Comment (smartalecH) [inline fmmax/fmm.py:34]: Would be nice to describe each of these in the docstring, along with an overall recommendation. That way these details appear in the online docs.

**Review by smartalecH — APPROVED** [review]
